grove/simon/simon.py

```diff
@@ -0,0 +1,450 @@
+    # can do by adding a vector with a single 1
+    # that can be inserted so that diag(W) is all ones
+    insert_row_num = 0
+    while insert_row_num < n - 1 and W[insert_row_num][insert_row_num] == 1:
+        insert_row_num += 1
+
+    new_row = np.zeros(shape=(n,))
+    new_row[insert_row_num] = 1
+    W = np.insert(W, insert_row_num, new_row, 0)
+
+    return W, insert_row_num
+
+
+def binary_back_substitute(W, s):
+    """
+    Perform back substitution on a binary system of equations.
+
+    Finds the :math:`\\mathbf{x}` such that
+    :math:`\\mathbf{\\mathit{W}}\\mathbf{x}=\\mathbf{s}`,
+    where all arithmetic is taken bitwise and modulo 2.
+
+    :param 2darray W: A square :math:`n\\times n` matrix of 0s and 1s,
+              in row-echelon form
+    :param list(int) s: An :math:`n\\times 1` vector of 0s and 1s
+    :return: The :math:`n\\times 1` vector of 0s and 1s that solves the above
+             system of equations.
+    :rtype: list(int)
+    """
+    # iterate backwards, starting from second to last row for back-substitution
+    s_copy = np.array(s)
+    n = len(s)
+    for row_num in xrange(n - 2, -1, -1):
+        row = W[row_num]
+        for col_num in xrange(row_num + 1, n):
+            if row[col_num] == 1:
+                s_copy[row_num] = int(s_copy[row_num]) ^ int(s_copy[col_num])
+
+    return s_copy
+
+
+def find_mask(cxn, oracle, qubits):
+    """
+    Runs Simon's algorithm to find the mask.
+
+    :param SyncConnection cxn: the connection used to run programs
+    :param Program oracle: the oracle to query;
+                   emulates a classical :math:`f(x)` function as a blackbox.
+    :param list(int) qubits: the input qubits
+    :return: a tuple t, where t[0] is the bitstring of the mask,
+             t[1] is the number of iterations that the quantum program was run,
+             and t[2] is said quantum program.
+    :rtype: tuple
+    """
+
+    # wrap the given oracle in the program to be used
+    simon_program = simon(oracle, qubits)
+    n = len(qubits)
+
+    # Generate n-1 linearly independent vectors
+    # that will be orthonormal to the mask s
+    # Done so by running the quantum program repeatedly
+    # and building up a row-echelon matrix W
+    iterations = 0
+    W = np.array([])
+    while True:
+        if len(W) == n - 1:
+            break
+        z = np.array(cxn.run_and_measure(simon_program, qubits)[0])
+        # attempt to insert z in such a way that
+        # W remains row-echelon
+        # and all rows are orthogonal to s
+        W = insert_into_row_echelon_binary_matrix(W, z)
+        iterations += 1
+
+    # make the matrix square by inserting a row
+    # that maintain W in row-echelon form
+    W, insert_row_num = make_square_row_echelon(W)
+
+    s = np.zeros(shape=(n,))
+    # inserted row is chosen not be orthogonal to s
+    s[insert_row_num] = 1
+
+    s = binary_back_substitute(W, s)
+
+    s_str = ''.join(map(str, map(int, s)))
```

Comment:
Do you need the conversion to `int` here? Isn't it a vector of `1`'s and `0`'s already? If so, can be simplified to: `''.join(str(x) for x in s)`